database.py

```diff
@@ -62,143 +67,276 @@ def as_picos(self) -> int:
 
     def __init__(self, *, auto_commit: bool = True) -> None:
         if Database.connection is None:
-            Database.connection = sqlite3.connect(config.settings.db.filename)
-            cur = Database.connection.cursor()
-            cur.execute(
-                """CREATE TABLE IF NOT EXISTS runs
-                (user TEXT, code TEXT, day INTEGER, part INTEGER, time REAL, answer INTEGER, answer2)"""
-            )
-            cur.execute(
-                """CREATE TABLE IF NOT EXISTS solutions
-                (key TEXT, day INTEGER, part INTEGER, answer INTEGER, answer2)"""
-            )
-            Database.connection.commit()
+            # assigns both to same object
+            con = Database.connection = sqlite3.connect(config.settings.db.filename)
+            cur = con.cursor()
+            _load_initial_schema(cur)
+            con.commit()
 
-        self._cursor = Database.connection.cursor()
-        self._auto_commit = auto_commit
+        self._cursor: sqlite3.Cursor = Database.connection.cursor()
+        self._auto_commit: bool = auto_commit
 
-    def __enter__(self) -> "Database":
+    def __enter__(self) -> Self:
         return self
 
     def __exit__(self, *ignore: object) -> None:
         if self._auto_commit:
             self._cursor.connection.commit()
 
-    # TODO(ultrabear): the current backend has no year concept, but this PR is for the API change first
-    def load_answers(self, _year: int, day: int, part: int, /) -> dict[str, str]:
-        """Load the expected answers for each input file."""
+    def load_answers(
+        self, year: int, day: AdventDay, part: AdventPart, /
+    ) -> dict[SessionLabel, str]:
+        """Load the expected answers for each input file, if they exist"""
+
+        # apply a literal here to avoid potential future runtime pollution
+        checkpart: Literal["answer_p1", "answer_p2"] = "answer_p1" if part == 1 else "answer_p2"
+
+        result = {
+            SessionLabel(label): str(ans)
+            for label, ans in self._cursor.execute(
+                f"SELECT session_label, {checkpart} FROM inputs WHERE (year = ? AND day = ? AND {checkpart} IS NOT NULL)",
+                (year, day),
+            )
+        }
+
+        return result
+
+    def save_submission(
+        self, author_id: int, year: int, day: AdventDay, part: AdventPart, code: bytes, /
+    ) -> SubmissionId:
+        """
+        Saves a benchmark submission to the database
+        this will insert a NULL into the average_time field,
+        it should be filled when all benchmarks are completed
+
+        Returns the unique SubmissionId that was generated for this submission
+        """
+
+        # TODO(ultrabear): assuming we are using the latest container
+        # version is probably ok maybe but its also kinda bad
+        # we should discuss this sometime idk
+        container_v = self._cursor.execute("SELECT MAX(id) FROM container_versions").fetchone()
```

Comment:
How will this table get populated?

There's probably two things we need to do differently here:

1. The containers need to have a metadata thingy that specifies what macro-version they are, in addition to the micro-version. We might want to do this by changing the image versioning scheme, not sure yet.
2. We probably need to have some kind of function/class/thingy that fetches the latest container tag every minute (or five or fifteen), and provides metadata based on that tag. This should work with major revisions, because the container _should_ be rebuilt before the new code is deployed, but I'd have to check the deployment flow before I commit to that.

Reply:
as for 2, this could probably be a cron job, or it could be an async sleeping task in process. It might make sense to just be an in process thing though, and also run on startup, since we only really care about container versions when the bot is actually running

Reply:
That's also an option. There's some subtle differences between the options that could impact correctness. I'll have to think about this some more.